### PR TITLE
feat: ungrouping empty directories

### PR DESCRIPTION
**Pull request by juefeiyan**
implement [#2646](https://github.com/nvim-tree/nvim-tree.lua/issues/2646)

change: if users enabled group_empty, there would be three states when users 'expand_or_collapse' a grouped folders:

1: closed grouped folder:
image

2: opened grouped folder:
image

3: opened ungrouped folder:
image

that feature would allow users to add files/folders in the middle grouped empty folder.

### CODE REVIEW

**Comment by alex-courtis** [comment]
This works:

File: lua/nvim-tree/lib.lua [hunk 1]
```diff
@@ -86,6 +87,32 @@ function M.get_last_group_node(node)
...
+  if M.group_empty and not is_root and child_folder_only then
+    node.group_next = child_folder_only
+    local ns = M.group_empty_folders(child_folder_only)
+    node.nodes = ns or {}
+    return ns
+  end
+  return node.nodes
+end
+
+---Group empty folders
```
Comment (alex-courtis) [inline lua/nvim-tree/lib.lua:105]: Group->ungroup and longer description please.

**Review by alex-courtis — CHANGES_REQUESTED** [review]
This is a change in behaviour that will interrupt current users. Also, group empty has historically been problematic and we need to reduce the blast radius.

Options:
1. Add a new option, defaulting off
2. Add API to specifically map a key to "expand the empties"

2 is preferable as it offers the greatest flexibility and ease of use.

@juefeiyan are you up to the challenge of 2?

File: lua/nvim-tree/lib.lua (hunk 1 above)
Comment (juefeiyan) [reply lua/nvim-tree/lib.lua:105]: sure,  will resolve the comments and working on option 2

Comment (juefeiyan) [reply lua/nvim-tree/lib.lua]: Just pushed a new version and implemented option 2, please take a look if you are free

**Comment by alex-courtis** [comment]
A test case like `mkdir a/b/C` functions as expected when the grouping is on and off.

File: lua/nvim-tree/api.lua [hunk 2]
```diff
@@ -175,12 +175,12 @@ end
...
-local function open_or_expand_or_dir_up(mode)
+local function open_or_expand_or_dir_up(mode, toggle_group)
   return function(node)
     if node.name == ".." then
       actions.root.change_dir.fn ".."
     elseif node.nodes then
-      lib.expand_or_collapse(node)
+      lib.expand_or_collapse(node, toggle_group)
     else
       edit(mode, node)
```
Comment (alex-courtis) [inline lua/nvim-tree/api.lua:185]: The other usages of this are for "open or edit" functionality.

This one is a toggle; falling back to edit is not intuitive. Let's do nothing on files please.

**Review by alex-courtis — CHANGES_REQUESTED** [review]
This looks great! It turned out better than I thought. I look forward to using it.

File: doc/nvim-tree-lua.txt [hunk 3]
```diff
@@ -2196,6 +2200,7 @@ You are encouraged to copy these to your own |nvim-tree.on_attach| function.
   vim.keymap.set('n', 'I',       api.tree.toggle_gitignore_filter,    opts('Toggle Filter: Git Ignore'))
   vim.keymap.set('n', 'J',       api.node.navigate.sibling.last,      opts('Last Sibling'))
   vim.keymap.set('n', 'K',       api.node.navigate.sibling.first,     opts('First Sibling'))
+  vim.keymap.set('n', 'L',       api.node.open.toggle_grouped_folders,opts('Toggle Grouped Folders'))
```
Comment (alex-courtis) [inline doc/nvim-tree-lua.txt:2203]:
```suggestion
  vim.keymap.set('n', 'G',       api.node.open.toggle_grouped_folders,opts('Toggle Group Empty'))
```

File: doc/nvim-tree-lua.txt [hunk 4]
```diff
@@ -1848,6 +1849,9 @@ node.open.vertical()                      *nvim-tree-api.node.open.vertical()*
 node.open.horizontal()                  *nvim-tree-api.node.open.horizontal()*
     |nvim-tree-api.node.edit()|, file will be opened in a new horizontal split.
 
+node.open.toggle_grouped_folders()      *nvim-tree-api.node.open.toggle_grouped_folders()*
```
[the remark this answers is not in the document]
Comment (juefeiyan) [reply doc/nvim-tree-lua.txt:1852]: thanks for the nice name

**Comment by juefeiyan** [comment]
> This looks great! It turned out better than I thought. I look forward to using it.
> 
> * [ ]  toggle_group_empty
> * [ ]  NOP on files
> * [ ]  mapping name and update-help.sh

Hi @alex-courtis , I jsut solved your comments, please take a look at my new MR

**Comment by alex-courtis** [comment]
This is looking great - files are indeed a NOP.

Sorry to be a pain; I have one more request:
Do not open folders, only toggle grouping. This function is very specifically a toggle.

I don't know if this is practical - please advise.

**Comment by juefeiyan** [comment]
> This is looking great - files are indeed a NOP.
> 
> Sorry to be a pain; I have one more request: Do not open folders, only toggle grouping. This function is very specifically a toggle.
> 
> I don't know if this is practical - please advise.

great suggestion, I just implement it, please take a look

**Review by alex-courtis — APPROVED** [review]
Works nicely. It's truly a toggle now and functions correctly even for never-opened folders.

Fair warning: group empty is historically fragile. I'll call on you if there's an unforseen issue and we may have to revert depending on the severity.

**Comment by przepompownia** [comment]
Thank you! I think I asked for a similar feature a long time ago, but I can't find any issue or discussion about it at the moment.

**Pull request merged**